[rocprofiler-sdk] Add Duration_ns field to kernel trace CSV output

ihhethan opened the pull request:
## Motivation

Adds missing Duration_ns field to kernel trace CSV output, addressing regression from rocprof to rocprofv3. Critical for analyzing applications where the same kernel has different durations per invocation.

## Technical Details

Modified Files:

- **rocpd/csv.py**
  - Added `Duration` column to kernel trace CSV query
  
- **conftest.py**
  - Defines pytest fixtures for JSON and rocpd DB inputs

- **validate.py**
  - Implements validation logic:
    - Extract kernel dispatch records from JSON
    - Query kernel data via rocpd Python API
    - Verify:
      - `Duration == End - Start`
      - Exact match between JSON and DB
      - All dispatch IDs align
      -
- **CMakeLists.txt**
  - Adds generate + validation test pipeline
  - Configures environment with `ROCPROF_OUTPUT_FORMAT=json,rocpd`

Change:

Before: "...End_Timestamp","LDS_Block_Size"...
After:  "...End_Timestamp","Duration_ns","LDS_Block_Size"...

## Test Plan



## Test Result

All locally tests passed on my machine
Duration_ns in column 12
Values verified: 80360 ns, 63895 ns, 54680 ns
Backward compatible

## Submission Checklist

- [ ] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

ihhethan committed:
Add Duration_ns field to kernel trace CSV output

Fixes: SWDEV-561821

This change adds the missing Duration_ns field to kernel trace CSV output,
addressing a regression from rocprof to rocprofv3.

Problem:
rocprofv3 --kernel-trace only provides timestamps for each kernel invocation,
unlike the CSV output from rocprof --stats which contained durations.
This is critical for applications like HemeLB where the same hotspot kernel
is called twice per iteration with different durations.

Solution:
Added Duration_ns field (calculated as end - start) to all trace CSV outputs.

Changes:
1. Python (rocpd/csv.py):
   - Added (end - start) AS Duration_ns to 6 trace functions:
     * write_kernel_csv
     * write_memory_copy_csv
     * write_memory_allocation_csv
     * write_counters_csv
     * write_scratch_memory_csv
     * write_region_csv
   - Ensured the column name uses the consistent "Duration_NS" spelling.

2. C++ (generateCSV.cpp):
   - Added \"Duration_ns\" to CSV header
   - Added (record.end_timestamp - record.start_timestamp) as the value
     for the Duration_NS column in kernel trace rows.

3. C++ (csv.hpp):
   - Updated kernel_trace_with_stream_csv_encoder from csv_encoder<22>
     to csv_encoder<23> to account for the new Duration_NS column.

4. Tests (rocprofv3 integration):
   - Added a new test under tests/rocprofv3/kernel-duration-ns:
     * rocprofv3-test-kernel-duration-ns-execute
       - Runs `rocprofv3 --kernel-trace -f csv` on the simple-transpose
         test application to generate out_kernel_trace.csv.
     * rocprofv3-test-kernel-duration-ns-validation
       - Runs validate.py with --csv-input=<out_kernel_trace.csv> and:
         - Asserts that Start_Timestamp, End_Timestamp, and Duration_NS
           columns exist in the header.
         - Verifies that Duration_NS == End_Timestamp - Start_Timestamp
           for multiple rows.

Testing:
- Environment: MI300X, ROCm 6.4.1, Docker container
- Test application: HPCTrainingExamples/daxpy_5
- Verified Duration_ns field appears in column 12
- Verified Duration_ns = End_Timestamp - Start_Timestamp
- Sample values: 80360 ns, 63895 ns, 54680 ns
- New tests:
  * ctest -R "kernel-duration-ns" -V
    - rocprofv3-test-kernel-duration-ns-execute
    - rocprofv3-test-kernel-duration-ns-validation
  * Both tests pass locally.

Impact:
- Restores functionality from legacy rocprof tool
- Enables efficient per-invocation kernel duration analysis
- No breaking changes to existing CSV format
- Backward compatible

SrirakshaNag commented on projects/rocprofiler-sdk/source/lib/output/generateCSV.cpp:
```diff
@@ -274,6 +274,7 @@ generate_csv(const output_config&
                                       "Correlation_Id",
                                       "Start_Timestamp",
                                       "End_Timestamp",
+                                      "Duration_ns",
```
The naming here needs to be consistent. Can it be Duration_NS?

ihhethan replied:
Sure

SrirakshaNag reviewed (dismissed):
How are we validating this new column in CSV?

Sushma1203 commented on projects/rocprofiler-sdk/source/lib/output/generateCSV.cpp:
```diff
@@ -313,6 +314,7 @@ generate_csv(const output_config&
                 record.correlation_id.internal,
                 record.start_timestamp,
                 record.end_timestamp,
+                (record.end_timestamp - record.start_timestamp),
```
Would it be possible to add a test case for this?

ihhethan replied:
I’ve added a new integration test under `tests/rocprofv3/kernel-duration-ns`:

- `rocprofv3-test-kernel-duration-ns-execute`
  - runs `rocprofv3 --kernel-trace -f csv` on `simple-transpose`
  - generates `out_kernel_trace.csv`

- `rocprofv3-test-kernel-duration-ns-validation`
  - runs `validate.py --csv-input <.../out_kernel_trace.csv>`
  - checks that the CSV header contains `Start_Timestamp`, `End_Timestamp`, and `Duration_NS`
  - verifies that `Duration_NS == End_Timestamp - Start_Timestamp` for multiple rows

Both tests pass locally:
ctest -R "kernel-duration-ns" -V

Sushma1203 commented on projects/rocprofiler-sdk/tests/rocprofv3/kernel-duration-ns/conftest.py:
```diff
@@ -0,0 +1,64 @@
+            rows.append(row)
+    if not rows:
+        raise RuntimeError(f"No data rows found in {csv_path}")
+    return rows
```
Can we move the validation of data to validate.py file instead of validating it here? 
Also it might be better to have a single fixture in conftest.py "csv_data" and follow the existing format present in other tests. Please refer [this](https://github.com/ROCm/rocm-systems/blob/develop/projects/rocprofiler-sdk/tests/rocprofv3/memory-allocation/conftest.py#L72)

ihhethan replied:
Done. Validation logic moved to validate.py and conftest.py now has single csv_data fixture following existing format. All tests passing.

Sushma1203 commented on projects/rocprofiler-sdk/tests/rocprofv3/kernel-duration-ns/validate.py:
```diff
@@ -0,0 +1,72 @@
+
+
+if __name__ == "__main__":
+    main()
```
Please refer to [this](https://github.com/ROCm/rocm-systems/blob/develop/projects/rocprofiler-sdk/tests/rocprofv3/memory-allocation/validate.py) format and update accordingly.

ihhethan committed:
SWDEV-561821: Refactor kernel duration validation tests per review feedback

- Consolidate to single csv_data fixture in conftest.py
- Move validation logic to validate.py
- Replace print statements with logging
- Maintain Duration_NS validation (Duration_NS = End_Timestamp - Start_Timestamp)

Sushma1203 commented on projects/rocprofiler-sdk/tests/rocprofv3/kernel-duration-ns/validate.py:
```diff
@@ -0,0 +1,134 @@
+            raise RuntimeError(
+                f"Row {i}: Duration_NS mismatch: got {duration}, expected {calc} "
+                f"(start={start}, end={end})"
+            )
```
@ihhethan can you also add a check to verify whether the start and end timestamps are > 0 and replace this with an assert instead of logging?

jrmadsen commented on projects/rocprofiler-sdk/source/lib/output/generateCSV.cpp:
```diff
@@ -260,31 +260,16 @@ generate_csv(const output_config&
+        tool::csv::kernel_trace_with_stream_csv_encoder{},
+        {"Kind",           "Agent_Id",         "Queue_Id",         "Stream_Id",
+         "Thread_Id",      "Dispatch_Id",      "Kernel_Id",        "Kernel_Name",
+         "Correlation_Id", "Start_Timestamp",  "End_Timestamp",    "Duration_NS",
```
We are not updating direct writing of CSV anymore. Please remove.

jrmadsen commented on projects/rocprofiler-sdk/source/lib/python/rocpd/csv.py:
```diff
@@ -235,6 +236,7 @@ def write_kernel_csv(importData, config) -> None:
         "stack_id AS Correlation_Id",
         "start AS Start_Timestamp",
         "end AS End_Timestamp",
+        "(end - start) AS Duration_NS",
```
```suggestion
        "(end - start) AS Duration",
```
Neither `Start_Timestamp` nor `End_Timestamp` have `_NS` suffixes even though they are also in nanoseconds. Be consistent.

jrmadsen commented on projects/rocprofiler-sdk/tests/rocprofv3/kernel-duration-ns/CMakeLists.txt:
```diff
@@ -0,0 +1,59 @@
+    NAME rocprofv3-test-kernel-duration-ns-execute
+    COMMAND
+        $<TARGET_FILE:rocprofiler-sdk::rocprofv3> -d
+        ${CMAKE_CURRENT_BINARY_DIR}/kernel-duration-ns -o out --output-format csv
```
Replace this test with one that generates JSON and rocpd output. We are not updating the legacy direct CSV output

jrmadsen commented on projects/rocprofiler-sdk/tests/rocprofv3/kernel-duration-ns/CMakeLists.txt:
```diff
@@ -0,0 +1,59 @@
+               FAIL_REGULAR_EXPRESSION
+               "${ROCPROFILER_DEFAULT_FAIL_REGEX}"
+               FIXTURES_REQUIRED
+               rocprofv3-test-kernel-duration-ns-run)
```
Use rocpd to convert to CSV and then compare with JSON output.

jrmadsen commented on projects/rocprofiler-sdk/tests/rocprofv3/kernel-trace-duration/validate.py:
You made changes to the rocpd CSV output and did zero validation on it.

ihhethan committed:
Merge branch 'develop' into user/ihhethan/SWDEV-561821

ihhethan committed:
Merge branch 'develop' into user/ihhethan/SWDEV-561821

ihhethan committed:
[SWDEV-561822] Add Duration column to rocpd kernel_trace CSV output

## Summary
Added 'Duration' field to kernel trace CSV output from rocpd to address
regression from legacy rocprof to rocprofv3. The Duration column provides
the kernel execution time in nanoseconds, calculated as (End_Timestamp - Start_Timestamp).

## Changes
- source/lib/output/csv.hpp: Added duration field to kernel dispatch CSV schema
- source/lib/output/generateCSV.cpp: Implemented duration calculation and output
- source/lib/python/rocpd/csv.py: Added Duration column to Python CSV generator
- tests/rocprofv3/kernel-duration-ns/*: Added comprehensive test suite

## Testing
- Verified Duration column exists in CSV output
- Validated Duration = End_Timestamp - Start_Timestamp
- Confirmed backward compatibility with existing CSV fields
- All integration tests passing

## Addresses
- SWDEV-561822: Request to add durations in --kernel-trace output
- Restores functionality from legacy rocprof --stats output

ihhethan committed:
Add Duration column validation test for kernel trace

- Generate JSON and rocpd output from same execution using ROCPROF_OUTPUT_FORMAT
- Convert rocpd database to CSV and validate Duration column
- Compare CSV Duration with JSON-derived duration (zero tolerance)
- Verify Duration = End_Timestamp - Start_Timestamp

Test validates:
- Duration column exists in CSV output
- Duration values exactly match between JSON and CSV
- Timestamps are identical (same execution source)
- Internal consistency of Duration calculation

Addresses: SWDEV-561822

ihhethan committed:
Merge branch 'develop' into user/ihhethan/SWDEV-561821

jrmadsen commented on projects/rocprofiler-sdk/tests/rocprofv3/kernel-duration-ns/validate.py:
```diff
@@ -0,0 +1,216 @@
+    os.makedirs(out_dir, exist_ok=True)
+    cmd = [sys.executable, "-m", "rocpd", "convert", "-i", db_path, "--output-format", "csv", "-d", out_dir]
+    res = subprocess.run(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)
+    assert res.returncode == 0, f"rocpd convert failed\ncmd={' '.join(cmd)}\nstdout={res.stdout}\nstderr={res.stderr}"
```
rocpd has a Python interface. Why would you run the command line in a subprocess when you can just call the API?

jrmadsen commented on projects/rocprofiler-sdk/tests/rocprofv3/kernel-duration-ns/validate.py:
```diff
@@ -0,0 +1,216 @@
+    for fn in os.listdir(out_dir):
+        if fn.endswith("kernel_trace.csv"):
+            return os.path.join(out_dir, fn)
+    assert False, f"kernel trace CSV not found in {out_dir}"
```
If you use the Python interface, you control where the file is written. The above approach is very error-prone.

jrmadsen commented on projects/rocprofiler-sdk/tests/rocprofv3/kernel-duration-ns/validate.py:
```diff
@@ -0,0 +1,216 @@
+        reader = csv.DictReader(f)
+        rows = list(reader)
+    assert len(rows) > 0, f"empty CSV: {path}"
+    return rows
```
Do you even need to write the CSV? I'd check with @yhuiYH ... you might just be able to call the CSV submodule and get the query that produces the data written to the CSV.

yhuiYH replied:
Unfortunately, CSV needs to be refactored to return the query.  As it is now, it will always write out a CSV file, you can't just get the query returned yet.

yhuiYH replied:
You could probably just update the new field to be checked in the `tests.integration.validate.rocprofv3-test-rocpd.test_csv_data` test : https://github.com/ROCm/rocm-systems/blob/develop/projects/rocprofiler-sdk/tests/rocprofv3/rocpd/CMakeLists.txt#L220

But I guess this `test_csv_data()` is called by both the rocpd-csv & generateCSV.cpp (legacy code) so it'll fail since the C++ code isn't going to add this duration field.
https://github.com/ROCm/rocm-systems/blob/develop/projects/rocprofiler-sdk/tests/pytest-packages/tests/rocprofv3.py#L356

jrmadsen commented on projects/rocprofiler-sdk/tests/rocprofv3/kernel-duration-ns/validate.py:
```diff
@@ -0,0 +1,216 @@
+        assert end >= start, f"end before start: start={start} end={end}"
+        
+        duration = end - start
+        m[str(dispatch_id)] = (start, end, duration)
```
Why `str(dispatch_id)`?

ihhethan committed:
tests: update kernel-duration validate to use rocpd Python API instead of CLI

ihhethan committed:
rocpd: add Duration column to kernel CSV output and validation test

Add Duration column computed as (end - start) in get_kernel_csv_query().
Add kernel-duration-ns integration test validating Duration against JSON output using rocpd Python API.

Validation ensures:
- Duration == End_Timestamp - Start_Timestamp
- rocpd DB values match JSON kernel records for the same execution.

yhuiYH commented on projects/rocprofiler-sdk/tests/rocprofv3/kernel-trace-duration/CMakeLists.txt:
```diff
@@ -0,0 +1,58 @@
+        -- $<TARGET_FILE:simple-transpose>)
+
+set_tests_properties(
+    rocprofv3-test-kernel-trace-duration-generate
```
I would prefer the test folder match the test name.  Please change your folder from `kernel-duration-ns` to `kernel-trace-duration`

yhuiYH commented on projects/rocprofiler-sdk/tests/rocprofv3/CMakeLists.txt:
```diff
@@ -48,8 +48,8 @@ add_subdirectory(conversion-script)
 add_subdirectory(python-bindings)
 add_subdirectory(rocpd)
 add_subdirectory(rocpd-kernel-rename)
+add_subdirectory(kernel-duration-ns)
```
```suggestion
add_subdirectory(kernel-trace-duration)
```

yhuiYH reviewed (commented):
1) Please fix all failing linting tests
2) Please change PR title to start with [rocprofiler-sdk]
3) Please update title of PR to be relevant to current change (drop the _ns)

ihhethan committed:
SWDEV-561821: rename kernel-duration-ns test to kernel-trace-duration and fix CI formatting

ihhethan reviewed (commented):
Hi, I’ve addressed all the review comments and pushed the updates.

Would you mind taking another look when you have time? Thanks!

yhuiYH commented on projects/rocprofiler-sdk/tests/rocprofv3/kernel-trace-duration/validate.py:
Why do you pass in repeated strings into `_get_first_present()`?  Seems a bit silly and error prone.

yhuiYH commented on projects/rocprofiler-sdk/tests/rocprofv3/kernel-trace-duration/validate.py:
Can't you just do a key.lower() to avoid passing in both strings?
I suggest you optimize this a bit.

ihhethan committed:
Merge branch 'develop' into user/ihhethan/SWDEV-561821

yhuiYH commented on projects/rocprofiler-sdk/tests/rocprofv3/kernel-trace-duration/validate.py:
```diff
@@ -0,0 +1,291 @@
+
+
+def _extract_dispatch_id_from_db_row(row):
+    value = _get_first_present(row, "dispatch_id", "correlation_id")
```
Do we really need to check both dispatch_id & correlation_id?  Seems like the `rocpd_kernel_dispatch` tables has constraint that `dispatch_id` is NOT NULL.  So it will always be populated: https://github.com/ROCm/rocm-systems/blob/develop/projects/rocprofiler-sdk/source/share/rocprofiler-sdk-rocpd/rocpd_tables.sql#L281

yhuiYH commented on projects/rocprofiler-sdk/tests/rocprofv3/kernel-trace-duration/validate.py:
```diff
@@ -0,0 +1,291 @@
+
+def _extract_duration_from_db_row(row):
+    value = _get_first_present(row, "duration")
+    return _as_int(value, field="duration")
```
```suggestion
column_names = ["dispatch_id", "start_timestamp", "end_timestamp", "duration"]

def _extract_values_from_db_row(row):
    return [
        _as_int(_get_first_present(row, name), field=f"{name}")
        for name in column_names
    ]
```
Your change is better than before, but can't we simplify it even better, like above?

ihhethan committed:
Merge branch 'develop' into user/ihhethan/SWDEV-561821

yhuiYH commented:
Changes look good now.
1) You should resolve any fixed suggestions/conversations in this PR to reduce the noise.
2) Unfortunately, the CI is in bad shape when you rebased.  You should try again after to ensure the tests are passing.  I would approve, but I want to see more passing tests first.

ihhethan committed:
Merge branch 'develop' into user/ihhethan/SWDEV-561821

ihhethan committed:
Merge branch 'develop' into user/ihhethan/SWDEV-561821

ihhethan committed:
Merge branch 'develop' into user/ihhethan/SWDEV-561821

ihhethan committed:
Merge branch 'develop' into user/ihhethan/SWDEV-561821

ihhethan committed:
Merge branch 'develop' into user/ihhethan/SWDEV-561821

ihhethan committed:
Merge branch 'develop' into user/ihhethan/SWDEV-561821

ihhethan committed:
Merge branch 'develop' into user/ihhethan/SWDEV-561821

yhuiYH reviewed (approved):
LGTM now that the CI tests are run and passing.  The sanitizer tests have not been running for a while I believe, so ignoring those pending tests.
